### PR TITLE
Navigation to authentication code page

### PR DESCRIPTION
**Pull request by mateuszc-kainos**
https://companieshouse.atlassian.net/browse/S4-32

image

### CODE REVIEW

File: services/infrastructure/mongo.initdb.js [hunk 1]
```diff
@@ -7,7 +7,8 @@ db.api_clients.insert({
   "client_secret" : "M2UwYzRkNzIwOGQ1OGQ0OWIzMzViYjJjOTEyYTc1",
   "user_id" : "Y2VkZWVlMzhlZWFjY2M4MzQ3MT",
   "redirect_uris" : [
-    "http://chs-dev/oauth2/user/callback"
+    "http://chs-dev/oauth2/user/callback",
+    "http://account.chs-dev/oauth2/user/callback"
```
Comment (peadarkelly) [inline services/infrastructure/mongo.initdb.js:11]: is this change still needed? I don't think so...

File: services/infrastructure/mongo.initdb.js [hunk 2]
```diff
@@ -24,6 +25,12 @@ db.api_clients.insert({
   "js_domains" : []
 })
 
+db.company_auth_codes.insert({
+    "_id" : "01777777",
+    "auth_code" : "$2a$10$uS7dsFz8iIuNvXQK6dG1v.F//uQajFz0BLc60/B8qrGqsdFrU77MO", // decrypted: 222222
+    "is_active" : true
```
Comment (peadarkelly) [inline services/infrastructure/mongo.initdb.js:31]: keep indentation consistent

File: src/inversify.config.ts [hunk 3]
```diff
@@ -31,18 +32,28 @@ export function initContainer(): Container {
...
     commonAuthMiddleware
   ))
+  const authConfig: AuthConfig = {
+    accountUrl: getEnvOrThrow('ACCOUNT_URL'),
+    accountRequestKey: getEnvOrThrow('ACCOUNT_OAUTH2_REQUEST_KEY'),
+    accountClientId: getEnvOrThrow('ACCOUNT_CH_CLIENT_ID'),
+    chsUrl: getEnvOrThrow('CHS_URL'),
+  }
+  container.bind(TYPES.CompanyAuthMiddleware).toConstantValue(
+    CompanyAuthMiddleware(cookieConfig, authConfig, createLogger(APP_NAME)))
```
Comment (peadarkelly) [inline src/inversify.config.ts:55]: No need to recreate a logger, just reuse the one defined here https://github.com/companieshouse/dissolution-web/pull/23/files#diff-06d803e02ba0c13e85350152879527f5R31

File: src/middleware/companyAuth.middleware.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import { generateNonce, jweEncodeWithNonce } from 'app/utils/jwt.encryption'
+
+export interface AuthConfig {
+  accountUrl: string,
+  accountRequestKey: string,
+  accountClientId: string,
+  chsUrl: string,
+}
+
+interface ISignInInfoWithCompanyNumber extends ISignInInfo {
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:19]: Seems like this should be added to the node session handler library as this is common functionality and not dissolutions specific

File: src/middleware/companyAuth.middleware.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  [SignInInfoKeys.CompanyNumber]?: string
+}
+
+const OATH_SCOPE_PREFIX = 'https://api.companieshouse.gov.uk/company/'
+
+export default function CompanyAuthMiddleware(config: CookieConfig, authConfig: AuthConfig,
+                                              logger: ApplicationLogger): RequestHandler {
+  return async (req: Request, res: Response, next: NextFunction) => {
+    const companyNumber = getCompanyNumberFromPath(req.originalUrl)
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:28]: It would be good to agree what format we will be retrieving the company number in:
1. Session
2. Query Params
3. Path params (I would avoid due to regex requirement)

File: src/middleware/companyAuth.middleware.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    if (companyNumber === '') {
+      return next(new Error('No Company Number in URL'))
+    }
+    const cookieId = req.cookies[config.cookieName]
+    if (cookieId) {
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:33]: I think the AuthMiddleware will handle this for us, so it's not needed in the Company Auth Middleware

File: src/middleware/companyAuth.middleware.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+  const found = path.match(regexPattern)
+  if (found) {
+    return found[1]
+  } else {
+    return ''
+  }
+}
+
+async function getAuthRedirectUri(req: Request, authConfig: AuthConfig, companyNumber?: string,): Promise<string> {
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:64]: watch the trailing comma

File: src/middleware/companyAuth.middleware.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  const originalUrl = req.originalUrl
+
+  let scope = ''
+
+  if (companyNumber != null) {
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:69]: i don't think companyNumber will ever be null or empty based off the check on L29.

File: src/middleware/companyAuth.middleware.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    scope = OATH_SCOPE_PREFIX + companyNumber
+  }
+  const nonce = generateNonce()
+
+  const mutableSession = req.session as Mutable<Session>
+  mutableSession.data[SessionKey.OAuth2Nonce] = nonce
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:75]: Will this work with setExtraData after the upgrade to V3?

File: src/middleware/companyAuth.middleware.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+async function createAuthUri(originalUri: string, nonce: string,
+                             authConfig: AuthConfig, scope?: string): Promise<string> {
+  let authUri = `${authConfig.accountUrl}/oauth2/authorise`.concat('?',
+    'client_id=', `${authConfig.accountClientId}`,
+    '&redirect_uri=', `${authConfig.chsUrl}/oauth2/user/callback`,
+    '&response_type=code')
+
+  if (scope != null) {
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:88]: scope will never be null

File: src/utils/jwt.encryption.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
+import { randomBytes } from 'crypto'
+import { JWE, JWK } from 'node-jose'
+
+import { AuthConfig } from 'app/middleware/companyAuth.middleware'
```
Comment (peadarkelly) [inline src/utils/jwt.encryption.ts:4]: This is a circular dependency. CompanyAuthMiddleware calls this file, and this file calls the CompanyAuthMiddleware file

File: src/utils/jwt.encryption.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+function generateNonce(): string {
+  const bytes = randomBytes(5)
+  const buffer = Buffer.from(bytes)
+  return buffer.toString('base64')
+}
+
+async function jweEncodeWithNonce(returnUri: string, nonce: string, authConfig: AuthConfig): Promise<string> {
+  const payloadObject: AuthPayload = {
+    'nonce': nonce,
+    'content': returnUri
```
Comment (peadarkelly) [inline src/utils/jwt.encryption.ts:25]: no need for quotes around keys.

File: src/middleware/companyAuth.middleware.ts (hunk 9 above)
Comment (mateuszc-kainos) [reply src/middleware/companyAuth.middleware.ts:71]: No. To accomplish this, a change in the Account service would need to be made.

File: src/inversify.config.ts [hunk 13]
```diff
@@ -10,7 +10,9 @@ import { authMiddleware as commonAuthMiddleware } from 'web-security-node'
 
 import { APP_NAME } from 'app/constants/app.const'
 import AuthMiddleware from 'app/middleware/auth.middleware'
+import CompanyAuthMiddleware, { AuthConfig } from 'app/middleware/companyAuth.middleware'
```
Comment (peadarkelly) [inline src/inversify.config.ts:13]: a default export with an additional export? That's an anti pattern

File: src/middleware/companyAuth.middleware.ts [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+  chsUrl: string,
+}
+
+const OATH_SCOPE_PREFIX = 'https://api.companieshouse.gov.uk/company/'
+
+export default function CompanyAuthMiddleware(authConfig: AuthConfig,
+                                              encryptionService: JwtEncryptionService,
+                                              logger: ApplicationLogger): RequestHandler {
+  return async (req: Request, res: Response, next: NextFunction) => {
+      const dissolutionSession: Optional<DissolutionSession> = req.session!.getExtraData(DISSOLUTION_SESSION_KEY)
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:27]: add some line breaks to this please - difficult to read when the whole function is squashed together

File: src/middleware/companyAuth.middleware.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+      if (!dissolutionSession?.companyNumber) {
+        return next(new Error('No Company Number in session'))
+      }
+      const companyNumber: Optional<string> = dissolutionSession!.companyNumber
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:31]: This shouldn't be optional as you have ensured it is present above. 
```
const companyNumber:string = dissolutionSession!.companyNumber!
```

File: src/middleware/companyAuth.middleware.ts [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+      const signInInfo: ISignInInfo = req.session!.get<ISignInInfo>(SessionKey.SignInInfo) || {}
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:32]: Do you need the default `|| {}`? Assuming the AuthMiddleware will ensure that this is definitely present

File: src/middleware/companyAuth.middleware.ts [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+      if (isAuthorisedForCompany(signInInfo, companyNumber)) {
+        logger.info(`User is authenticated for ${companyNumber}`)
+        return next()
+      } else {
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:36]: you probably don't need the else here as the if block returns:
```
if (isAuthorisedForCompany(signInInfo, companyNumber)) {
  return next()
}

return res.redirect(...)
```

File: src/middleware/companyAuth.middleware.ts [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+  return signInInfo[SignInInfoKeys.CompanyNumber] === companyNumber
+}
+
+async function getAuthRedirectUri(req: Request, authConfig: AuthConfig,
+                                  encryptionService: JwtEncryptionService, companyNumber?: string): Promise<string> {
+  const originalUrl = req.originalUrl
+
+  const scope = OATH_SCOPE_PREFIX + companyNumber
+
+  const nonce = encryptionService.generateNonce()
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:54]: add types here - i assume this is a string?

File: src/middleware/companyAuth.middleware.ts [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+  const encodedNonce = await encryptionService.jweEncodeWithNonce(originalUrl, nonce)
+
+  const mutableSession = req.session as Mutable<Session>
+  mutableSession.data[SessionKey.OAuth2Nonce] = nonce
+  req.session = mutableSession as Session
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:59]: Can this be done in a session service using:
```
req.session.setExtraData(SessionKey.OAuth2Nonce, nonce)
```

File: src/middleware/companyAuth.middleware.ts [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+                             authConfig: AuthConfig, scope: string): Promise<string> {
+  let authUri = `${authConfig.accountUrl}/oauth2/authorise`.concat('?',
+    'client_id=', `${authConfig.accountClientId}`,
+    '&redirect_uri=', `${authConfig.chsUrl}/oauth2/user/callback`,
+    '&response_type=code')
+
+  authUri = authUri
+    .concat('&scope=', scope)
+    .concat('&state=', encodedNonce)
+  return authUri
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:74]: I think this can all be concated at once and use template strings throughout:
```
return `${authConfig.accountUrl}/oauth2/authorise`.concat(
    '?',
    `client_id=${authConfig.accountClientId}`,
    `&redirect_uri=${authConfig.chsUrl}/oauth2/user/callback`,
    `&response_type=code',
    `&scope=${scope}`,
    `&state =${encodedNonce}`
)
```

File: test/middleware/companyAuth.middleware.test.ts [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+    req.session!.getExtraData = getExtraDataStub
+
+    const res = {} as Response
+    const redirectStub: sinon.SinonStub = sinon.stub()
+    res.redirect = redirectStub
+
+    const next = sinon.stub()
+
+    await middleware(req, res, next)
+    assert.isTrue(redirectStub.calledOnce)
```
Comment (peadarkelly) [inline test/middleware/companyAuth.middleware.test.ts:92]: can you assert parts of the URL to maximise coverage? The nonce is the only thing you can't guarantee:
```
await middleware(req, res, next)
assert.isTrue(redirectStub.calledOnce)

const redirectUrl: string = redirectStub.args[0][0]

assert.include(redirectUrl, 'http://account.chs-dev/oauth2/authorise?client_id=something&...')
```

File: test/middleware/companyAuth.middleware.test.ts [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+  let authConfig: AuthConfig
+
+  const COMPANY_NUMBER = '12345678'
+
+  beforeEach(() => {
+    encryptionService = mock(JwtEncryptionService)
+    logger = mock(ApplicationLogger)
+
+    authConfig = {
+      chsUrl: '',
```
Comment (peadarkelly) [inline test/middleware/companyAuth.middleware.test.ts:33]: seems like we should add some default values here so that we can assert parts of the redirectUrl

File: src/middleware/companyAuth.middleware.ts (hunk 19 above)
Comment (mateuszc-kainos) [reply src/middleware/companyAuth.middleware.ts:59]: No, because account doesn't look inside ExtraData for the nonce key

Comment (peadarkelly) [reply src/middleware/companyAuth.middleware.ts:59]: No problem, can we keep all req.session in the sessions service though please?

**Comment by mateuszc-kainos** [comment]
> Can you also add the middleware to the selectDirector.controller.ts ?

Added

File: src/middleware/companyAuth.middleware.ts [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+                                  sessionService: SessionService,
+                                  companyNumber?: string): Promise<string> {
+  const originalUrl: string = req.originalUrl
+  const scope: string = OATH_SCOPE_PREFIX + companyNumber
+  const nonce: string = encryptionService.generateNonce()
+  const encodedNonce: string = await encryptionService.jweEncodeWithNonce(originalUrl, nonce)
+
+  const mutableSession = sessionService.getSession(req) as Mutable<Session>
+  mutableSession.data[SessionKey.OAuth2Nonce] = nonce
+  sessionService.setSession(req, mutableSession as Session)
```
Comment (peadarkelly) [inline src/middleware/companyAuth.middleware.ts:56]: This is fine, but just for context what I was suggesting was:
```
sessionService.setCompanyAuthNonce(nonce)
```
so that the middleware would just execute one line

File: src/services/encryption/jwtEncryption.service.ts [hunk 24]
```diff
@@ -0,0 +1,52 @@
+import 'reflect-metadata'
+
+import { randomBytes } from 'crypto'
+import { provide } from 'inversify-binding-decorators'
+import { JWE, JWK } from 'node-jose'
+
+import AuthConfig from 'app/models/authConfig'
+
+
```
Comment (peadarkelly) [inline src/services/encryption/jwtEncryption.service.ts:9]: remove one of these lines plz

File: src/services/encryption/jwtEncryption.service.ts [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+interface AuthPayload {
+  nonce: string,
+  content: string
+}
+
+@provide(JwtEncryptionService)
+export class JwtEncryptionService {
```
Comment (peadarkelly) [inline src/services/encryption/jwtEncryption.service.ts:16]: could this be default?

File: test/middleware/companyAuth.middleware.test.ts [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+    sessionService = mock(SessionService)
+    logger = mock(ApplicationLogger)
+
+    when(sessionService.getSession(anything())).thenReturn(
+      {
+        data:
+          {
+            [SessionKey.OAuth2Nonce]: ''
+          }
+      } as Session)
```
Comment (peadarkelly) [inline test/middleware/companyAuth.middleware.test.ts:41]: this feels like it should be added to session.fixtures.ts
